과제3 MVP 예제 변경

dlwls5201 opened the pull request:
1. 데이터 레이어에 에러 핸들링 관련 로직 추가
2. MVP로 변경
3. 테스트 가능하도록 일부 코드 수정
4. Search와 Repo 각각 PresenterTest 코드 작성

dlwls5201 committed:
에러 핸들링 기능 추가

dlwls5201 committed:
ErrorEntity : sealed class -> open class 변경

dlwls5201 committed:
Add Resource Providers

dlwls5201 committed:
make Base Presenter and BaseView

dlwls5201 committed:
change RepositoryActivity to MVP

dlwls5201 committed:
change Item for test

dlwls5201 committed:
make RepositoryPresenterTest code

dlwls5201 committed:
update searchActivity to MVP

dlwls5201 committed:
delete clearItems

dlwls5201 committed:
add SearchPresenterTest code

dlwls5201 committed:
move hideKeyboard to top

dlwls5201 reviewed (commented):
Room까지 작업 후 라밸 붙이겠습니다.

dlwls5201 committed:
import room library

dlwls5201 committed:
add room code in domain module

dlwls5201 committed:
add room code in data module

dlwls5201 committed:
add room code in app module

dlwls5201 reviewed (commented):
MainActivity 기능 적용
- 클릭한 레포 내부 디비에 저장 후 불러오기
- 전체 레포 삭제 버튼 추가

> RoomDatabase를 data module에서 생성하고 싶었지만 room-runtime 라이브러리를 불러오질 못해서 app module에서 RoomDatabase 선언 및 초기화를 해주었습니다.

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/di/LocalModule.kt:
```diff
@@ -0,0 +1,31 @@
+            "repo.db"
+        ).build()
+    }
+}
```
data module에 apply plugin: 'kotlin-android'를 추가하고 manifest에 package를 지정해주면 사용할 수 있습니다.

dlwls5201 replied:
명준쓰 떙큐! 적용 완료
91e73dd

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/presentation/main/MainActivity.kt:
```diff
@@ -1,21 +1,107 @@
+
+    override fun showClearRepoHistory() {
+        val empty = Collections.emptyList<RepoItem>()
+        searchAdapter.setItems(empty)
```
searchAdapter.setItems(emptyList()) 이렇게 한줄로도 가능할 것 같습니다

dlwls5201 replied:
수정했습니다.
3c0c6ad

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/presentation/main/MainActivity.kt:
```diff
@@ -1,21 +1,107 @@
 
+        btnActivityMainClear.setOnClickListener {
+            alert(message = "데이터를 모두 지웁니다.") {
+                positiveButton("확인") {
```
귀찮겠지만.. string resource로..

dlwls5201 replied:
수정했습니다.
bb179fb

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/presentation/main/MainActivity.kt:
```diff
@@ -1,21 +1,107 @@
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
 
+        presenter = MainPresenter(
+            this,
+            getUsecase,
+            clearUsecase,
+            resourcesProvider
+        )
```
presenter도 koin을 이용해서 주입 받는건 어떨까요?

dlwls5201 replied:
이...이건 다음에..MVVM 때 할래 :)

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/model/UserItem.kt:
```diff
@@ -12,9 +11,9 @@ data class UserItem(
 )
 
-fun UserEntity.mapToPresentation(resources: Resources) = let {
+fun UserEntity.mapToPresentation(resources: ResourcesProvider) = let {
```
let 대신 run을 사용해보는건 어떨까요?
let과 return 값은 동일하지만 블록 내 argument가 this여서 it을 사용하지 않아도 될 것 같습니다.

dlwls5201 replied:
수정했습니다.
6357487

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/di/ResourcesProvider.kt:
```diff
@@ -0,0 +1,15 @@
+import androidx.annotation.PluralsRes
+import androidx.annotation.StringRes
+
+interface ResourcesProvider {
```
ResourcesProvider는 어떤 의도로 사용하셨나요?

dlwls5201 replied:
모듈화 및 테스트를 위해 사용했습니다.
1. app 모듈 외에 data, domain 모듈에서는 AndroidScheulders.main 을 선언할 수가 없어 이를 app 모듈에서 선언해 추상화로 넘길러고 했습니다.

2.  AndroidScheulders.main처럼 Android 의존성 있는 코드가 있는경우 테스트가 진행되지 않습니다. 테스트 시 TestScheulderPreovider에 Schedulers.trampoline()선언한 것처럼 Android 의존성을 제거해 테스트 하기 위함입니다.

dlwls5201 replied:
적용예시
41ef4a0

myung6024 replied:
혹시 참고한 코드가 있나요?

dlwls5201 replied:
범준쓰 코드
https://github.com/omjoonkim/GitHubBrowserApp

myung6024 commented on dlwls5201/MashUpArchitecture/app/src/main/java/com/tistory/mashuparchitecture/di/ResourcesProvider.kt:
```diff
@@ -0,0 +1,15 @@
+        quantity1: Int,
+        quantity2: Int
+    ): String?
+}
```
ResourcesProvider와 ResourcesProviderImpl은 util 디렉토리에 넣고 di 디렉토리는 di와 관련된 파일만 넣는게 좋지 않을까요?

dlwls5201 replied:
위치 변경
f9ef101

dlwls5201 committed:
Room 초기화 위치 Data 모듈로 변경
참고 링크(https://github.com/bufferapp/clean-architecture-components-boilerplate/blob/09f746223d/presentation/build.gradle)

dlwls5201 committed:
emptyList()로 간편 초기화

dlwls5201 committed:
string resource 추기 및 적용

dlwls5201 committed:
let -> run 변경

dlwls5201 committed:
GetRepoUsecaseTest 코드 추가

dlwls5201 committed:
Provider 패키지 위치 변경

myung6024 reviewed (approved):
LGTM bbb
